File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3

let parseOperation (arg : string) =
    match arg with
    | "+" -> Calculator.CalculatorOperation.Plus
    | "-" -> Calculator.CalculatorOperation.Minus
    | "*" -> Calculator.CalculatorOperation.Multiply
    | "/" -> Calculator.CalculatorOperation.Divide
    | _ -> Calculator.CalculatorOperation.Undefined
      
let parseCalcArguments(args : string[]) =
    match isArgLengthSupported args with
    | false -> ArgumentException() |> raise
    | true ->
        match parseOperation args[1] with
        | CalculatorOperation.Undefined -> ArgumentException() |> raise
        | _ ->
            match Double.TryParse args[0], Double.TryParse args[2] with
            | (true, res1), (true, res2) -> {arg1 = Double.Parse args[0]
                                             arg2 = Double.Parse args[2]
                                             operation = parseOperation args[1]}
            | _ -> ArgumentException() |> raise
            
            
            
            
            
            
            
    
    

File: Homework8/Hw8/Parser/Parser.cs
using System.Globalization;
using Hw8.Calculator;

namespace Hw8.Parser;

public static class Parser
{
    public static ParseResult ParseCalcArguments(string[] args)
    {
        if (!double.TryParse(args[0], NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out var val1) ||
            !double.TryParse(args[2], NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out var val2))
            return new ParseResult(0, Operation.Invalid, 0, Messages.InvalidNumberMessage);
        var operation = ParseOperation(args[1]);
        return operation switch
        {
            Operation.Invalid => new ParseResult(val1, operation, val2, Messages.InvalidOperationMessage),
            Operation.Divide when val2 < double.Epsilon => new ParseResult(0, Operation.Invalid, 0,
                Messages.DivisionByZeroMessage),
            _ => new ParseResult(val1, operation, val2)
        };
    }
    
    private static Operation ParseOperation(string arg)
    {
        return arg.ToLower() switch
        {
            "plus" => Operation.Plus,
            "minus" => Operation.Minus,
            "multiply" => Operation.Multiply,
            "divide" => Operation.Divide,
            _ => Operation.Invalid
        };
    }
}